### PR TITLE
Return code differences in wolfSSL_EVP_PKEY_cmp et al.

### PR DESCRIPTION
**Pull request by TakayukiMatsuo**
This PR is to fix issues reported in  ZD11899.
Return code differences in the following functions are addressed:
- EVP_CIPHER_CTX_cleanup
- BIO_free
- EVP_PKEY_cmp

### CODE REVIEW

File: wolfcrypt/src/evp.c [hunk 1]
```diff
@@ -1975,17 +1975,29 @@ WOLFSSL_API int wolfSSL_EVP_PKEY_missing_parameters(WOLFSSL_EVP_PKEY *pkey)
...
+/* wolfSSL_EVP_PKEY_cmp
+ * returns 0 on success, -1 on failure.
+ *
+ * This behavior is different from openssl.
+ *  EVP_PKEY_cmp returns:
+ *    1 : two keys match
+ *    0 : do not match
+ *    -1: key types are different
+ *    -2: the operation is not supported
+ * If you mant this function behave the same as openSSL,
```
Comment (cconlon) [inline wolfcrypt/src/evp.c:1987]: Spelling - "mant" should be "want"

File: wolfcrypt/src/evp.c [hunk 2]
```diff
@@ -1975,17 +1975,29 @@ WOLFSSL_API int wolfSSL_EVP_PKEY_missing_parameters(WOLFSSL_EVP_PKEY *pkey)
...
+ * define WOLFSSL_ERROR_CODE_OPENSSL so that WS_RETURN_CODE fills the gap.
```
Comment (cconlon) [inline wolfcrypt/src/evp.c:1988]: Let's change the wording of "fills the gap" to "translates return codes to match OpenSSL equivalent behavior".

File: wolfcrypt/src/evp.c [hunk 3]
```diff
@@ -4124,9 +4139,10 @@ int wolfSSL_EVP_MD_type(const WOLFSSL_EVP_MD *md)
             }
             ctx->gcmAuthInSz = 0;
 #endif
+            return WOLFSSL_SUCCESS;
         }
 
-        return WOLFSSL_SUCCESS;
+        return WOLFSSL_FAILURE;
```
Comment (cconlon) [inline wolfcrypt/src/evp.c:4145]: OpenSSL man page for EVP_CIPHER_CTX_cleanup() states that it returns 1 on success.  Looking at the OpenSSL 1.1.1j source code EVP_CIPHER_CTX_cleanup() maps to EVP_CIPHER_CTX_reset(), which returns 1 if the input "ctx" is NULL.  This means we would still return WOLFSSL_SUCCESS (1) here.

If this is different than expected, maybe it varies with version of OpenSSL?

Comment (TakayukiMatsuo) [reply wolfcrypt/src/evp.c]: openSSL 1.0.2 returns "1" also. I revert this change.

File: wolfcrypt/src/evp.c (hunk 2 above)
Comment (TakayukiMatsuo) [reply wolfcrypt/src/evp.c]: Changed the wording.

File: wolfcrypt/src/evp.c (hunk 1 above)
Comment (TakayukiMatsuo) [reply wolfcrypt/src/evp.c]: Fixed it.

**Review by cconlon — APPROVED** [review]

**Comment by cconlon** [comment]
Assigning to Todd for final review / merge.

File: src/ssl.c [hunk 4]
```diff
@@ -16286,8 +16286,9 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
             }
 
             XFREE(bio, 0, DYNAMIC_TYPE_OPENSSL);
+            return WOLFSSL_SUCCESS;
         }
-        return 1;
+        return WOLFSSL_FAILURE;
```
Comment (toddouska) [inline src/ssl.c:16291]: Are you sure this is correct?  Why is bio == NULL a failure?

Comment (TakayukiMatsuo) [reply src/ssl.c:16291]: openSSL1.1.1 returns 0 as following:
 
```
int BIO_free(BIO *a)
{
    int ret;

    if (a == NULL)
        return 0;

```

**Review by toddouska — APPROVED** [review]

**Pull request merged**